Packages/com.unity.inputsystem/InputSystem/API/Input.cs

```diff
@@ -0,0 +1,1862 @@
+
+            InitializeGlobalActions(defaultGlobalActionsPath, globalActionsAssetPath);
+        }
+
+        internal static void Shutdown()
+        {
+            if (s_PointerAction != null)
+            {
+                s_PointerAction.performed -= OnPointerMoved;
+                s_PointerAction.Disable();
+                s_PointerAction = null;
+            }
+
+            if (s_ScrollAction != null)
+            {
+                s_ScrollAction.performed -= OnMouseScrolled;
+                s_ScrollAction.Disable();
+                s_ScrollAction = null;
+            }
+
+            InputSystem.onDeviceChange -= OnDeviceChange;
+
+            ShutdownGlobalActions();
+        }
+
+        private static void OnPointerMoved(InputAction.CallbackContext context)
+        {
+            s_PointerPosition = context.ReadValue<Vector2>();
+        }
+
+        private static void OnMouseScrolled(InputAction.CallbackContext context)
+        {
+            s_ScrollDelta = context.ReadValue<Vector2>();
+        }
+
+        private static void OnDeviceChange(InputDevice device, InputDeviceChange change)
+        {
+            if (change == InputDeviceChange.Added)
+            {
+                AddGamepadToFirstFreeSlot(device);
+                AddJoystickToFirstFreeSlot(device);
+            }
+            else if (change == InputDeviceChange.Removed || change == InputDeviceChange.Disconnected)
+            {
+                RemoveGamepad(device);
+                RemoveJoystick(device);
+            }
+        }
+
+        private static void AddGamepadToFirstFreeSlot(InputDevice device)
+        {
+            if (!(device is Gamepad gamepad)) return;
+
+            for (var i = 0; i < maxGamepadSlots; i++)
+            {
+                if (s_Gamepads[i] != null) continue;
+
+                var frameCount = Time.frameCount;
+#if UNITY_EDITOR
+                if (!s_TimeHasUpdatedThisFrame)
+                    frameCount += 1;
+#endif
+                s_Gamepads[i] = gamepad;
+                s_GamepadsConnectedFrames[i] = frameCount;
+                break;
+            }
+        }
+
+        private static void RemoveGamepad(InputDevice device)
+        {
+            for (var i = 0; i < maxGamepadSlots; i++)
+            {
+                if (s_Gamepads[i] != device) continue;
+
+                s_Gamepads[i] = null;
+
+                var frameCount = Time.frameCount;
+#if UNITY_EDITOR
+                if (!s_TimeHasUpdatedThisFrame)
+                    frameCount += 1;
+#endif
+                s_GamepadsDisconnectedFrames[i] = frameCount;
+                break;
+            }
+        }
+
+        private static void AddJoystickToFirstFreeSlot(InputDevice device)
+        {
+            if (!(device is Joystick joystick)) return;
+
+            for (var i = 0; i < (int)InputSlot.Joystick_Max; i++)
+            {
+                if (s_Joysticks[i] != null) continue;
+
+                s_Joysticks[i] = joystick;
```

Comment:
This basic add/remove to free slot feels like it only preserves the joystick index for simple single device add/remove and would fail to preserve index if I remove one device, add a different one, then add my old device back.